Bind9 DNSSEC (#69)
canonical/ubuntu-server-documentation: how-to/networking/install-dnssec.md

(install-dnssec)=
# Installing DNS Security Extensions (DNSSEC)

DNSSEC is a set of security extensions to DNS which allow DNS data to be verified for authenticity and integrity.

> This how-to will show how to enable DNSSEC for an existing zone in your bind9 DNS server deployment.

Reviewer: ```suggestion
This guide will show you how to enable DNSSEC for an existing zone in your bind9 DNS server deployment.
```

Reviewer: As above - should bind9 be BIND9?

## Starting point

> The starting point for this how-to is an existing bind9 DNS server deployed with an authoritative zone. For details on how to deploy bind9 in this fashion, please see the [DNS How-To](https://github.com/panlinux/ubuntu-server-documentation/blob/bind9-dnssec/how-to/networking/install-dns.md). One key difference from that guide, however, is that we need the zone file to be in a directory where bind9 can write to, like `/var/lib/bind`, instead of `/etc/bind`.

Reviewer: ```suggestion
The starting point for this how-to is an existing bind9 DNS server deployed with an authoritative zone. For details on how to deploy bind9 in this fashion, please see the {ref}`DNS How-To <install-dns>`. One key difference from that guide, however, is that we need the zone file to be in a directory where bind9 can write to, like `/var/lib/bind`, instead of `/etc/bind`.
```

First, install the bind9 package:

    sudo apt install bind9 -y

Edit `/etc/bind/named.conf.local` and add this *zone* block:

    zone "example.internal" {
        type master;
        file "/var/lib/bind/db.example.internal";
    };

Create the file `/var/lib/bind/db.example.internal` with these contents:

    $TTL 86400      ; 1 day
    example.internal.       IN SOA  example.internal. root.example.internal. (
                                    1          ; serial
                                    43200      ; refresh (12 hours)
                                    900        ; retry (15 minutes)
                                    1814400    ; expire (3 weeks)
                                    7200       ; minimum (2 hours)
                                    )
    example.internal.       IN  NS      ns.example.internal.
    ns                      IN  A       192.168.1.10
    noble                   IN  A       192.168.1.11

Restart the service:

    sudo systemctl restart named

Check if the service can resolve the name `noble.example.internal`:

    $ dig @127.0.0.1 +short noble.example.internal
    192.168.1.11

## Enabling DNSSEC

> Enabling DNSSEC for a zone involves multiple steps. Thankfully for us, the bind9 DNS server takes care of all of them by default, automatically, leaving very little for us to do. Converting a zone in this way means at least generating new keys, and signing all the resource records of the zone. But that is only "day 1": such a zone must be maintained properly. Keys must be rotated, expiration dates must be set, etc. The current versions of bind9 take care of that with a dnssec policy, and of course there is a default one that can be used from the get go.

Reviewer: Will the user know which keys we mean when we say "generating new keys" and "rotating keys"?

Author: The idea here is to just show how much more complicated things could be, how there are many important details pertaining DNSSEC, but that these are taken care of by the server via the dnssec policy.

> Enabling DNSSEC for a zone involves multiple steps. Thankfully for us, the bind9 DNS server takes care of all of them by default, automatically, leaving very little for us to do. Converting a zone in this way means at least generating new keys, and signing all the resource records of the zone. But that is only "day 1": such a zone must be maintained properly. Keys must be rotated, expiration dates must be set, etc. The current versions of bind9 take care of that with a dnssec policy, and of course there is a default one that can be used from the get go.

Reviewer: ```suggestion
Enabling DNSSEC for a zone involves multiple steps. Thankfully for us, the bind9 DNS server takes care of all of them by default, automatically, leaving very little for us to do. Converting a zone in this way means at least generating new keys, and signing all the resource records of the zone. But that is only "day 1": such a zone must be maintained properly. Keys must be rotated, expiration dates must be set, etc. The current versions of bind9 take care of that with a DNSSEC policy, and of course there is a default one that can be used from the start.
```

    named[3246]: zone example.internal/IN (signed): next key event: 23-Oct-2024 22:47:12.544
    named[3246]: any newly configured zones are now loaded
    named[3246]: running
    named[3246]: resolver priming query complete: success
    named[3246]: managed-keys-zone: Key 20326 for zone . is now trusted (acceptance timer complete)
    named[3246]: zone example.internal/IN (signed): sending notifies (serial 3)

Depending on the zone size, signing all records can take longer.

A few interesting events can be seen in the logs above:

  * Keys were generated for the *example.internal* zone.
  * The *example.internal* zone became signed.
  * A *key event* was scheduled. This is bind9 also taking care of the maintenance tasks of the signed zone and its keys.
  * Since the zone changed, its serial number was incremented (started as 1, now it's 3).

The DNSSEC keys are kept in `/var/cache/bind`:

    -rw-r--r-- 1 bind bind  413 Oct 23 20:50 Kexample.internal.+013+48112.key
    -rw------- 1 bind bind  215 Oct 23 20:50 Kexample.internal.+013+48112.private
    -rw-r--r-- 1 bind bind  647 Oct 23 20:50 Kexample.internal.+013+48112.state

> This is the bulk of the work. This zone is now signed, and maintained automatically by bind9 using the *default* dnssec policy.

Reviewer: ```suggestion
This is the bulk of the work. This zone is now signed, and maintained automatically by bind9 using the *default* DNSSEC policy.
```

## Verification

The zone that was just signed is almost ready to serve DNSSEC. Let's perform some verification steps.

> As it is now, this zone *example.internal* is "disconnected" from the parent zone. Its name was made up for this how-to, but even if it represented a realm domain, it would still be missing the connection to the parent zone. Remember that DNSSEC relies on the chain of trust, and the parent of our zone needs to be able to vouch for it.

Reviewer: Should we link here to the explanation page?

Author: I think it's a bit late in the howto/guide for that, and would look odd to link to the explanation for the first time here, more than halfway through it.

Before taking that step, however, it's important to verify if everything else is working. In particular, we would want to perform some DNSSEC queries, and perform validation. A good tool to perform this validation is `delv`.

`delv` is similar to `dig`, but it will perform validation on the results using the same internal resolver and validator logic as the bind9 server itself.

Since our zone is disconnected, we need to tell `delv` to use the public key created for the zone as a trusted anchor, and to not try to reach out to the root servers of the internet.

First, copy the public zone key somewhere else so it can be edited. For example:

    cp /var/cache/bind/Kexample.internal.+013+48112.key /tmp/example.key

> That file will have some comments and the top, and then have a line that starts with the zone name, like this (the full key was truncated below for brevity):

Reviewer: ```suggestion
That file will have some comments at the top, and then have a line that starts with the zone name, like this (the full key was truncated below for brevity):
```

    $ dnssec-dsfromkey /var/cache/bind/Kexample.internal.+013+48112.key
    example.internal. IN DS 48112 13 2 1212DE7DA534556F1E11898F2C7A66736D5107962A19A0BFE1C2A67D6841962A

### DNSKEY record format
The DNSKEY format doesn't need tooling, and can be found directly in the zone's public key file, after the lines that start with the comment delimiter `;`:

    $ cat /var/cache/bind/Kexample.internal.+013+48112.key
    ; This is a key-signing key, keyid 48112, for example.internal.
    ; Created: 20241023205023 (Wed Oct 23 20:50:23 2024)
    ; Publish: 20241023205023 (Wed Oct 23 20:50:23 2024)
    ; Activate: 20241023205023 (Wed Oct 23 20:50:23 2024)
    ; SyncPublish: 20241024215523 (Thu Oct 24 21:55:23 2024)
    example.internal. 3600 IN DNSKEY 257 3 13 jkmS5hfyY3nSww4tD9Fy5d+GGc3A/zR1CFUxmN8T2TKTkgGWp8dusllM 7TrIZTEg6wZxmMs754/ftoTA6jmM1g==

Taking that line, the actual record to send to the parent zone just needs a little bit of tweaking so it looks like this:

    example.internal. 3600 IN DNSKEY 257 3 13 (jkmS5hfyY3nSww4tD9Fy5d+GGc3A/zR1CFUxmN8T2TKTkgGWp8dusllM 7TrIZTEg6wZxmMs754/ftoTA6jmM1g==);

## Further reading

>  * The DNSSEC Guide from bind9: https://bind9.readthedocs.io/en/stable/dnssec-guide.html
>  * Easy-Start Guide for Signing Authoritative Zones: https://bind9.readthedocs.io/en/stable/dnssec-guide.html#signing
>  * Creating a Custom DNSSEC Policy: https://bind9.readthedocs.io/en/stable/dnssec-guide.html#signing-custom-policy
>  * Detailed DNSSEC chapter from the bind9 documentation: https://bind9.readthedocs.io/en/stable/chapter5.html
>  * `delv` manual page: https://manpages.ubuntu.com/manpages/noble/man1/delv.1.html
>  * Working with the parent zone: https://bind9.readthedocs.io/en/latest/dnssec-guide.html#working-with-the-parent-zone

Reviewer: ```suggestion
 * [The DNSSEC Guide from bind9](https://bind9.readthedocs.io/en/stable/dnssec-guide.html)
 * [Easy-Start Guide for Signing Authoritative Zones](https://bind9.readthedocs.io/en/stable/dnssec-guide.html#signing)
 * [Creating a Custom DNSSEC Policy](https://bind9.readthedocs.io/en/stable/dnssec-guide.html#signing-custom-policy)
 * [Detailed DNSSEC chapter from the bind9 documentation](https://bind9.readthedocs.io/en/stable/chapter5.html)
 * [`delv` manual page](https://manpages.ubuntu.com/manpages/noble/man1/delv.1.html)
 * [Working with the parent zone](https://bind9.readthedocs.io/en/latest/dnssec-guide.html#working-with-the-parent-zone)
```
